components/remote_settings/src/storage.rs

```diff
@@ -19,37 +22,75 @@ use crate::{Attachment, RemoteSettingsRecord, Result};
-///     the same time.  Maybe we should create a single write connection and put it behind a mutex
-///     to avoid the possibility of SQLITE_BUSY.  Or maybe not, the writes seem like they should be
-///     very fast.
-///   - Maybe we should refactor this to use the DAO pattern like suggest does.
-pub struct Storage {}
+pub struct Storage {
+    conn: Arc<Mutex<Connection>>,
+}
 
 impl Storage {
-    pub fn new(_path: Utf8PathBuf) -> Result<Self> {
-        Ok(Self {})
+    pub fn new(path: Utf8PathBuf) -> Result<Self> {
+        let conn = Connection::open(path)?;
+        let storage = Self {
+            conn: Arc::new(Mutex::new(conn)),
+        };
+        storage.initialize_database()?;
+
+        Ok(storage)
+    }
+
+    // Create the different tables for records and attachements for every new sqlite path
+    fn initialize_database(&self) -> Result<()> {
+        let conn = self.conn.lock().unwrap();
+        conn.execute_batch(
+            "
+            CREATE TABLE IF NOT EXISTS records (
+                id TEXT PRIMARY KEY,
+                collection_url TEXT NOT NULL,
+                data BLOB NOT NULL
+            );
+            CREATE TABLE IF NOT EXISTS attachments (
+                id TEXT PRIMARY KEY,
+                collection_url TEXT NOT NULL,
+                data BLOB NOT NULL
+            );
+            ",
+        )?;
+        Ok(())
     }
 
     /// Get the last modified timestamp for the stored records
     ///
     /// Returns None if no records are stored or if `collection_url` does not match the
     /// `collection_url` passed to `set_records`.
-    pub fn get_last_modified_timestamp(&self, _collection_url: &str) -> Result<Option<u64>> {
-        Ok(None)
+    pub fn get_last_modified_timestamp(&self, collection_url: &str) -> Result<Option<u64>> {
+        let conn = self.conn.lock().unwrap();
+        let mut stmt = conn.prepare(
+            "SELECT MAX(json_extract(data, '$.last_modified')) FROM records WHERE collection_url = ?",
+        )?;
+
+        let result: Option<u64> = stmt.query_row(params![collection_url], |row| row.get(0))?;
+
+        Ok(result)
     }
 
     /// Get cached records for this collection
     ///
-    /// Returns None if no records are stored or if `collection_url` does not match the
-    /// `collection_url` passed to `set_records`.
-    pub fn get_records(&self, _collection_url: &str) -> Result<Option<Vec<RemoteSettingsRecord>>> {
-        Ok(None)
+    /// Returns None if no records are stored or if `collection_url` does not match the `collection_url` passed
+    /// to `set_records`.
+    pub fn get_records(&self, collection_url: &str) -> Result<Option<Vec<RemoteSettingsRecord>>> {
+        let conn = self.conn.lock().unwrap();
+        let mut stmt = conn.prepare("SELECT data FROM records WHERE collection_url = ?")?;
+        let mut rows = stmt.query(params![collection_url])?;
+        let mut records = Vec::new();
+        while let Some(row) = rows.next()? {
```

Comment:
I usually prefer [query_and_then](https://github.com/mozilla/application-services/blob/d160a87f30a8c1c752d8323f7b42f8ad46a6b450/components/suggest/src/db.rs#L265-L270) for this kind of iteration.  I'm not sure if it's better or worse in this case, I just wanted to point it out to you.